composer/utils/data.py

```diff
@@ -93,27 +97,86 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
     return image_tensor, target_tensor
 
 
-def add_dataset_transform(dataset, transform):
-    """Flexibly add a transform to the dataset's collection of transforms.
+def add_dataset_transform(dataset: torch.utils.data.Dataset,
+                          transform: Callable,
+                          location: Union[str, type] = "end",
+                          pre_post: str = "pre") -> Dataset:
+    """Flexibly add a transform to the dataset's collection of transforms. Warning: this
+    function will not behave as expected if a dataset's collection of transforms contains
+    nested torchvision.transforms.Compose (e.g dataset.transforms =
+    transforms.Compose([transform0, transform1, transforms.Compose(...)])). This function
+    should be updated to flatten nested compositions.
 
     Args:
         dataset: A torchvision-like dataset
         transform: Function to be added to the dataset's collection of transforms
+        location (str, type), optional: Where to insert the transform in
+            the sequence of transforms. "end" will append to the end or "start" will
+            insert at index 0, a transform (e.g. torchvision.transforms.ToTensor) will
+            insert pre- or post- the first instance of `location`, determined by the
+            `pre_post` argument. Default = "end". If `location` is not in list of
+            transforms, will append to end.
+        pre_post (str), optional: Whether to insert before ("pre") or after ("post") the
+            transform string passed in 'location'. Ignored if transform = "start" or "end". Default = "pre".
 
     Returns:
         The original dataset. The transform is added in-place.
     """
 
     if not hasattr(dataset, "transform"):
-        raise ValueError(f"Dataset of type {type(dataset)} has no attribute 'transform'. Expected TorchVision dataset.")
+        raise AttributeError(textwrap.dedent(f"""Dataset must have attribute `transform`."""))
+
+    valid_non_transform_locations = ["start", "end"]
+    valid_pre_post_values = ["pre", "post"]
+    if location not in valid_non_transform_locations and pre_post not in valid_pre_post_values:
+        raise ValueError(
+            textwrap.dedent(
+                f"Invalid value combination for argument `pre_post`: `{pre_post} and `location`: {location}. If location is not one of ['start', 'end'], `pre_post` must be one of ['pre', 'post']."
+            ))
+    if not hasattr(dataset, "transform"):
+        raise AttributeError(textwrap.dedent(f"Argument `dataset` must have attribute `transform`."))
+
+    added_transform = 0
+
+    def added_transform_warning():
+        log.warning(
+            f"Transform {transform} added to dataset. Dataset now has the following transforms: {dataset.transform}")
```

Comment:
How do I deal with all the 
```
Cannot access member "transform" for type "Dataset[Unknown]"
Member "transform" is unknown
```
errors from the static analysis?

Reply:
Hmm guess the static typechecker doesn't understand the guard you have upfront.

I'm not a typing expert, but I think you would use `Protocol` to implement a new type that has the attribute `transform`, and use that instead of `torch.utils.Dataset`, e.g. see [PEP 544](https://www.python.org/dev/peps/pep-0544/). 

But maybe defer to @ravi-mosaicml on this one.

Reply:
Can you add an `assert isinstance(dataset, VisionDataset)`?